### PR TITLE
flux-queue: add idle subcommand

### PR DESCRIPTION
**Pull request by garlick**
This PR provides a command `flux queue idle`, similar to `flux queue drain`, that waits until there are no pending alloc requests to the scheduler, and no jobs in RUN or CLEANUP state.  One can now cleanly terminate running work when shutting down an instance with

```console
$ flux queue stop
$ flux job cancelall -f --states=RUN
$ flux queue idle
```

The `stop` prevents the job manager from sending new alloc requests to the scheduler, and sends cancel messages for any pending ones.  The `cancelall` terminates running work.  The `idle` then blocks until the canceled jobs have transitioned to INACTIVE state (with eventlogs etc committed to the KVS).

I added this sequence to rc3 (speculatively).

As discussed in #2649, in a system instance this could be preceded with something like
```
$ flux queue stop
$ flux job raiseall -s1 --states=RUN checkpoint
$ sleep 600
```
which would allow jobs some time to respond to the `checkpoint` exception, including submitting new work`, before either terminating normally on their own or getting canceled in rc3.

Possibly we'll want to repackage this a `flux shutdown` or similar but for now, just getting some basic mechanisms ironed out.

Tests are still needed, but wanted to post this for early feedback.

### CODE REVIEW

**Comment by grondo** [comment]
This looks really useful to me! I think the interface is fine. My only slight quibble is that `flux queue idle` sounds more like a command to idle the queue along the lines of `stop` and `disable`, rather than a command that blocks until some state.

However, my only other suggestion would be to add a `wait` verb in there, e.g. `flux queue wait-idle`, which I don't really like much more, or have a `flux queue wait STATE` where STATE would be idle in this case. If we took the second approach, then `flux queue drain` might also become `flux queue wait drained`, which I really don't like. :man_shrugging: 

In summary, I think it looks good as is!

**Comment by garlick** [comment]
OK coverage is decent now, and it sounds like there are no better ideas for the command naming (which I agree is a bit unsatisfying, oh well).   Ready for a review.

**Comment by grondo** [comment]
There is still a race with the use of `flux job list` to check the count of submitted jobs:

```
expecting success: 
	COUNT=$(flux job list | wc -l) &&
	test ${COUNT} -eq 3

not ok 39 - flux-queue: there are 3 active jobs
#	
#		COUNT=$(flux job list | wc -l) &&
#		test ${COUNT} -eq 3
#	

```

**Comment by garlick** [comment]
Oops!  I meant to use the job manager one.  Fixed and forced a push.

**Comment by grondo** [comment]
In rc3, we may want to run the `flux queue` and `flux jobs` commands in a `--quiet` mode, or redirect stderr.
```console
(flux-25099-1gmRRZ) grondo@fluke108:~/git/f.git$ exit
exit
2020-02-06T18:44:59.464741Z broker.err[0]: rc3: flux-queue: Scheduling is disabled
2020-02-06T18:44:59.495114Z broker.err[0]: rc3: flux-job: Command matched 0 jobs
```

**Comment by garlick** [comment]
Good point.

Just pushed a couple of commits. LMK how that looks and I'll squash the incremental development if good.

**Comment by grondo** [comment]
> Just pushed a couple of commits. LMK how that looks and I'll squash the incremental development if good.

Looks good!

Not to focus on minutiae here, but it might also be nice if there was a message when the instance exits with some pending jobs, e.g. something that looks like this
```diff
diff --git a/etc/rc3 b/etc/rc3
index 64e77cd..e13f84d 100755
--- a/etc/rc3
+++ b/etc/rc3
@@ -3,6 +3,10 @@
 flux queue stop --quiet
 flux job cancelall --quiet -f --states RUN
 flux queue idle
+njobs=$(flux job stats | jq '.job_states.depend + .job_states.sched')
+if test $njobs -gt 0; then
+    printf "Exiting with %d pending jobs\n" $njobs >&2
+fi
 
 core_dir=$(cd ${0%/*} && pwd -P)
 all_dirs=$core_dir${FLUX_RC_EXTRA:+":$FLUX_RC_EXTRA"}
``` 

(not that I advocate  including `jq` in our rc scripts though!)


Then you can play fun games like:
```console
t$ src/cmd/flux start sh -c 't/ingest/submitbench -r 16 job.json'
6123683840
6140461056
6140461057
6140461058
6140461059
6140461060
6140461061
6157238272
6157238273
6157238274
6157238275
6157238276
6174015488
6174015489
6174015490
6174015491
2020-02-06T21:35:10.761120Z broker.err[0]: rc3: flux-job: Canceled 1 jobs (0 errors)
2020-02-06T21:35:10.840689Z broker.err[0]: rc3: Exiting with 12 pending jobs
```

**Comment by garlick** [comment]
Hah cool.  Ok, I added a pending job count to the idle response.  Now you get:
```
2020-02-06T22:34:06.246615Z broker.err[0]: rc3: flux-job: Canceled 2 jobs (0 errors)
2020-02-06T22:34:06.304055Z broker.err[0]: rc3: flux-queue: 8 pending jobs
```
How's that?

(flux queue idle has a `--quiet` option to suppress the pending count if zero)

**Comment by grondo** [comment]
> How's that?

Nice! :+1:

BTW, the new `rc3` is really helpful! No more good for nothing "orphan" jobs laying around.

**Comment by grondo** [comment]
asan builder failed with something I haven't seen before:
```
PASS: t2204-job-info.t 34 - flux job list outputs ntasks correctly
2020-02-06T23:10:40.920029Z job-info.err[0]: info_lookup_continuation: flux_kvs_lookup_get: Is a directory
2020-02-06T23:10:40.920348Z job-info.err[0]: main_namespace_lookup_continuation: flux_rpc_get_unpack: Is a directory
2020-02-06T23:11:13.461702Z job-info.err[0]: lookup_cb: flux_request_unpack: Protocol error
2020-02-06T23:11:13.553283Z job-info.err[0]: watch_cb: flux_request_unpack: Protocol error
2020-02-06T23:11:13.641983Z job-info.err[0]: guest_watch_cb: flux_request_unpack: Protocol error
2020-02-06T23:11:15.273062Z broker.err[0]: Run level 2 Exited with non-zero status (rc=1) 178.9s
2020-02-06T23:11:15.558846Z broker.err[0]: rc3: flux-queue: Scheduling is disabled
2020-02-06T23:11:15.821630Z broker.err[0]: rc3: flux-job: Command matched 0 jobs
2020-02-06T23:11:16.070793Z broker.err[0]: rc3: flux-queue: 0 pending jobs
flux-start: 0 (pid 10824) exited with rc=1
 (4 tasks)
ok 35 - reload the job-info module
PASS: t2204-job-info.t 35 - reload the job-info module
ok 36 - verify task count preserved across restart
PASS: t2204-job-info.t 36 - verify task count preserved across restart
not ok 37 - flux job list outputs nnodes/ranks correctly (1 task / 1 node)
FAIL: t2204-job-info.t 37 - flux job list outputs nnodes/ranks correctly (1 task / 1 node)
#	
#	        jobid=`flux mini submit -n1 hostname` &&
#	        echo $jobid > nodecount1.id &&
#	        flux job wait-event $jobid clean >/dev/null &&
#	        obj=$(flux job list -s inactive | grep $jobid) &&
#	        echo $obj | jq -e ".nnodes == 1" &&
#	        echo $obj | jq -e ".ranks == \"0\""
#	
ok 38 - flux job list outputs nnodes/ranks correctly (2 tasks, / 1 node)
PASS: t2204-job-info.t 38 - flux job list outputs nnodes/ranks correctly (2 tasks, / 1 node)
ok 39 - flux job list outputs nnodes/ranks correctly (3 tasks, / 2 nodes)
PASS: t2204-job-info.t 39 - flux job list outputs nnodes/ranks correctly (3 tasks, / 2 nodes)
ok 40 - flux job list outputs nnodes/ranks correctly (5 tasks, / 3 nodes)
PASS: t2204-job-info.t 40 - flux job list outputs nnodes/ranks correctly (5 tasks, / 3 nodes)
ok 41 - reload the job-info module
PASS: t2204-job-info.t 41 - reload the job-info module
not ok 42 - verify nnodes preserved across restart
FAIL: t2204-job-info.t 42 - verify nnodes preserved across restart
#	
#	        jobid1=`cat nodecount1.id` &&
#	        jobid2=`cat nodecount2.id` &&
#	        jobid3=`cat nodecount3.id` &&
#	        jobid4=`cat nodecount4.id` &&
#	        obj=$(flux job list -s inactive | grep ${jobid1}) &&
#	        echo $obj | jq -e ".nnodes == 1" &&
#	        echo $obj | jq -e ".ranks == \"0\"" &&
#	        obj=$(flux job list -s inactive | grep ${jobid2}) &&
#	        echo $obj | jq -e ".nnodes == 1" &&
#	        echo $obj | jq -e ".ranks == \"0\"" &&
#	        obj=$(flux job list -s inactive | grep ${jobid3}) &&
#	        echo $obj | jq -e ".nnodes == 2" &&
#	        echo $obj | jq -e ".ranks == \"[0-1]\"" &&
#	        obj=$(flux job list -s inactive | grep ${jobid4}) &&
#	        echo $obj | jq -e ".nnodes == 3" &&
#	        echo $obj | jq -e ".ranks == \"[0-2]\""
#	
ok 43 - list count / max_entries works
PASS: t2204-job-info.t 43 - list count / max_entries works
ok 44 - list request with empty attrs works
PASS: t2204-job-info.t 44 - list request with empty attrs works
```

**Comment by grondo** [comment]
I think the failure above is most likely due to a new bug introduced by 931d9e7456cbf7d241815f7168deb09c2c5f6443. Unrelated to the work here so I'll restart the builder and hope we don't hit the issue again.
